### ISSUE TITLE
Make use of Glow OpenGL binding generator.

### ISSUE THREAD

**Issue by slimsag** [azul3d/engine#25]
_From @slimsag on August 11, 2014 5:31_

Derived from #5.

Azul3D's existing tool for generating Go OpenGL wrappers glwrap (https://github.com/azul3d/cmd-glwrap) and Glow (https://github.com/go-gl/glow) have a lot in common. It makes sense to combine efforts.

Glow features a better generator, with actual support for more GL extensions (glwrap doesn't really). So we will adopt the use of Glow in the long run. Before we can adopt Glow, we need to ensure it can live up to (at least) the performance of glwrap.

glwrap supports Go->C function call batching, Glow does not. There are no benchmarks that tell us what kind of performance gain (or loss) this call batching feature actually gives us. We need to:

glwrap supports debug builds, which checks with glError() at each OpenGL function call. If any error exists an Go panic occurs and the previous ~200 GL function calls and their arguments are logged.

* Benchmark call batching and see if we should implement it in Glow.
* (Not needed in order to switch, but would be extremely useful): Implement debug builds in Glow.
* (Not needed in order to switch, but would be very useful): Make Glow only generate bindings to functions/definitions we actually need.

_Copied from original issue: azul3d/oldissues#17_

**Comment by slimsag** [comment]
_From @dskinner on August 11, 2014 6:59_

I'm not sure how much a benchmark would really have to tell for if batching is useful. As I understand it, opengl funcs return immediately which means nearly all the overhead of making a call is going to be in cgo.

Below is a rudimentary test just to measure overhead of cgo. On my system, best case has 47ns of overhead. Worse case, 100ns of overhead.

Looking at only the best case with only 20K calls per frame, that would roughly equal 1ms of time spent in overhead. Assuming 60 FPS then that's roughly 6% of the runtime in cgo overhead.

```
// go run -gcflags '-N -l' ./main.go
package main

/*
static int foo() {
	return 42;
}
*/
import "C"

import (
	"fmt"
	"time"
)

func foo() int { return 42 }

const N = 100000

func main() {
	var (
		t  time.Time
		dt time.Duration
	)

	t = time.Now()
	for n := 0; n < N; n++ {
		foo()
	}
	dt = time.Since(t)
	fmt.Printf("go:  %vns\n", dt.Nanoseconds()/N)

	t = time.Now()
	for n := 0; n < N; n++ {
		C.foo()
	}
	dt = time.Since(t)
	fmt.Printf("cgo: %vns\n", dt.Nanoseconds()/N)
}
```

**Comment by slimsag** [comment]
(@errcw, @shurcooL may be interested in this)

I've finally taken the time to write a proper benchmark for CGO call batching. I believe this effectively measures the performance difference between https://github.com/azul3d/cmd-glwrap and https://github.com/go-gl/glow as far as call batching goes.

Given the benchmark statistics listed over at https://github.com/slimsag/cgo-batching, I would say that the performance gain using call batching is large enough to warrant additional thought.

I'll be working in the interim to see how hard it would be to add this to Glow; but at least one other concern remains: If Go pointers cannot land in C space due to new GC changes, it would severely break this and the work for be for nothing.

Stephen

**Comment by slimsag** [comment]
_From @errcw on August 16, 2014 3:1_

Thanks for doing the detailed analysis.

> I'll be working in the interim to see how hard it would be to add this to Glow

I spent a bit of time thinking about the easiest ways to make this work and my glib conclusion is that it is a real shame that C does not support closures. Realistically I think we'd end up with a structure that mirrors that of azul. Given the extensive changes we could create a new `package.tmpl` file for batch vs. online.

> but at least one other concern remains: If Go pointers cannot land in C space due to new GC changes, it would severely break this and the work for be for nothing

If and when these GC changes land I expect all Go OpenGL bindings to be broken in some way. With or without batching I anticipate needing a significant restructuring.

**Comment by slimsag** [comment]
I created an issue over at go-gl/glow#40 so we can track progress there.

> If and when these GC changes land I expect all Go OpenGL bindings to be broken in some way. With or without batching I anticipate needing a significant restructuring.

I agree completely. It may make more sense for me to work on this once more information is known regarding what GC changes might occur, lest this work be for nothing.

**Comment by slimsag** [comment]
Closed in azul3d/gfx-gl2#11

**Comment by slimsag** [comment]
Not sure why this remained open, was fixed long ago.
